controlplane/kubeadm/internal/workload_cluster_test.go

```diff
@@ -487,6 +488,117 @@ imageRepository: k8s.gcr.io
 	}
 }
 
+func TestUpdateApiServerInKubeadmConfigMap(t *testing.T) {
+	validAPIServerConfig := `apiServer:
+  certSANs:
+  - foo
+  extraArgs:
+    foo: bar
+  extraVolumes:
+  - hostPath: /foo/bar
+    mountPath: /bar/baz
+    name: mount1
+  timeoutForControlPlane: 3m0s
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+`
+	kubeadmConfig := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kubeadmConfigKey,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Data: map[string]string{
+			clusterConfigurationKey: validAPIServerConfig,
+		},
+	}
+
+	kubeadmConfigNoKey := kubeadmConfig.DeepCopy()
+	delete(kubeadmConfigNoKey.Data, clusterConfigurationKey)
+
+	kubeadmConfigBadData := kubeadmConfig.DeepCopy()
+	kubeadmConfigBadData.Data[clusterConfigurationKey] = `badconfig`
+
+	g := NewWithT(t)
+	scheme := runtime.NewScheme()
+	g.Expect(corev1.AddToScheme(scheme)).To(Succeed())
+	tests := []struct {
+		name              string
+		apiServer         kubeadmv1beta1.APIServer
+		objs              []client.Object
+		expectErr         bool
+		expectedAPIServer string
+	}{
+		{
+			name:      "updates the config map",
+			apiServer: kubeadmv1beta1.APIServer{CertSANs: []string{"foo", "bar"}},
+			objs:      []client.Object{kubeadmConfig},
+			expectErr: false,
+			expectedAPIServer: `apiServer:
+  certSANs:
+  - foo
+  - bar
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+`,
+		},
+		{
+			name:              "returns error if cannot find config map",
+			expectErr:         true,
+			expectedAPIServer: validAPIServerConfig,
+		},
+		{
+			name:              "returns error if config has bad data",
+			objs:              []client.Object{kubeadmConfigBadData},
+			apiServer:         kubeadmv1beta1.APIServer{CertSANs: []string{"foo", "bar"}},
+			expectErr:         true,
+			expectedAPIServer: validAPIServerConfig,
+		},
+		{
+			name:              "returns error if config doesn't have cluster config key",
+			objs:              []client.Object{kubeadmConfigNoKey},
+			apiServer:         kubeadmv1beta1.APIServer{CertSANs: []string{"foo", "bar"}},
+			expectErr:         true,
+			expectedAPIServer: validAPIServerConfig,
+		},
+		{
+			name: "should not update config map if no changes are detected",
```

Comment:
Wanted to check if `client.Update` is not called here, but couldn't find a way to do this with fakeClient.

Reply:
What about using testenv instead if the fake client, see https://cluster-api.sigs.k8s.io/developer/testing.html#unit-tests for more info about our current stance on testing

Reply:
I haven't tried it yet, was just following the previous test conventions. Will try it today, and see it how it goes.

Reply:
I've replaced `fakeClient` with `testenv`, and I'm using `ResourceVersion` to check if the configmap was updated or not. wdyt?